### DISCUSSION TITLE
Using multiplexed buttons for menu navigation?

### DISCUSSION THREAD

**Discussion by MiqViq** [Spirik/GEM#89] [General]
I am designing a MIDI foot controller for my own use, based on ESP32-C3/S3.
Display will be either 128x64 SPI LCD (preferred) or a 16x2/20x4 I2C LCD.

I need some kind of menu managing system for setting up global parameters and editing preset names and values.
After initial testing GEM looks quite promising.

I have already 12 buttons multiplexed using 16ch analog mpx for selecting presets and sending CC/PC commands and it is working fine.
As the device will have multiple buttons, RGB LED channel, SPI and I2C devices the available pins are getting low...
Just started browsing the GEM docs but I did not find any clue how to use multiplexed buttons as controllers for menus.
Is it even possible with current GEM version?

**Comment by MiqViq** [comment]
If I understood correctly I can send menu navigation commands directly to GEM from the main code:
https://github.com/Spirik/GEM/wiki/How-to#control-menu-from-your-sketch
So instead initializing the display with this:
`lcdScreen.begin(/*Select/OK=*/ 15, /*Right/Next=*/ 25, /*Left/Prev=*/ 26, /*Up=*/ 32, /*Down=*/ 33, /*Home/Cancel=*/ 17);`
I will use this for display init:
`lcdScreen.begin();`

Can I then keep my logic for handling the buttons and send action request directly to active GEM menu items?

I will post some results after I have done my own testing on this.

**Comment by Spirik** [comment]
Sure, controlling GEM with multiplexed signals is absolutely possible! I've done it myself in my personal projects. I used [KeyDetector](https://github.com/Spirik/KeyDetector) library for that instead of built-in U8g2 key detection (btw, KeyDetector is used throughout GEM examples for Adafruit GFX and AltSerialGraphicLCD versions).

But of course it doesn't mean that you _should_ use KeyDetector for multiplexed key detection, since, as I understand, you've already have your own implementation of key multiplex/detection. If that is the case, than you should be able to control GEM from your sketch by calling `menu.registerKeyPress()` with corresponding key identifier. Refer to [Control menu from your sketch](https://github.com/Spirik/GEM/wiki/How-to#control-menu-from-your-sketch) section in wiki for correct key identifiers, and also check out this [example](https://github.com/Spirik/GEM/issues/57) in which GEM is controlled through Serial.

(And here is an [example](https://github.com/Spirik/KeyDetector/wiki/Example-03%3A-Multiplexed-signal-readings) of how to work with multiplexed signals in KeyDetector, although in that particular case without integration with GEM. But generally speaking, example on detection level of [analog signals](https://github.com/Spirik/KeyDetector/wiki/Example-01:-Analog-signal-readings) is may be even more relevant - KeyDetector can listen to the specific levels of analog signal and trigger different actions accordingly).
